### PR TITLE
[VCDA-3312] Support for proxy configuration

### PR DESCRIPTION
**Pull request by sakthisunda**
- VCDCluster level proxy support added
- containerd proxy variables added to clustertypes
- inject proxy variables thru machine controller
- TBD: testing
- Verified proxy variables placed in the http-proxy.conf of control-plane vm
- Manually verified wget on driver files goes thru proxy
- Tested with new capvcd cluster creation with control plane and worker node

image

Squid access log:

image

---
This change is [Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143)

### CODE REVIEW

File: api/v1beta1/vcdcluster_types.go [hunk 1]
```diff
@@ -76,6 +83,8 @@ type VCDClusterSpec struct {
 	ParentUID string `json:"parentUid,omitempty"`
 	// +optional
 	UseAsManagementCluster bool `json:"useAsManagementCluster,omitempty"`
+	// +optional
```
Comment (arunmk) [inline api/v1beta1/vcdcluster_types.go:86]: there is a work from Yan I think which tries to keep status in sync with spec. So we need similar fields in status as well. @sahithi is that correct?

File: controllers/cluster_scripts/cloud_init_control_plane.yaml [hunk 2]
```diff
@@ -31,6 +31,16 @@ write_files:
       storageProfile: "%s"
       filesystem: "%s"
     ---
+- path: /etc/environment
```
Comment (arunmk) [inline controllers/cluster_scripts/cloud_init_control_plane.yaml:34]: We can skip this file and directly put the `%s` in the `/root/control_plane.sh` below, correct?

File: controllers/cluster_scripts/cloud_init_control_plane.yaml [hunk 3]
```diff
@@ -49,6 +59,7 @@ write_files:
        fi
        vmtoolsd --cmd "info-set guestinfo.post_customization_cloud_init_output $cloud_init_output"
     }
+    source /etc/environment
```
Comment (arunmk) [inline controllers/cluster_scripts/cloud_init_control_plane.yaml:62]: no need of this as mentioned above

**Review by arunmk — APPROVED** [review]
:lgtm: with some comments.

**Review by arunmk — COMMENTED** [review]
This needs to be done for the worker node as well.

File: controllers/cluster_scripts/cloud_init_control_plane.yaml (hunk 2 above)
Comment (sakthisunda) [reply controllers/cluster_scripts/cloud_init_control_plane.yaml]: Done

**Review by sahithi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143)* status: 0 of 5 files reviewed, 5 unresolved discussions (waiting on @arunmk, @sahithi, and @sakthisunda)

___
*[`api/v1alpha4/zz_generated.conversion.go` line 111 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-N1uGaeHESjbR_c8nLoq:-N1uGaeHESjbR_c8nLor:b6ecyus) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/6da227c8a28bd4efed245e17d13cc7d4477a1269/api/v1alpha4/zz_generated.conversion.go#L111)):*
> ```Go
> 	}
> 	if err := s.AddGeneratedConversionFunc((*v1beta1.VCDMachineStatus)(nil), (*VCDMachineStatus)(nil), func(a, b interface{}, scope conversion.Scope) error {
> 		return Convert_v1beta1_VCDMachineStatus_To_v1alpha4_VCDMachineStatus(a.(*v1beta1.VCDMachineStatus), b.(*VCDMachineStatus), scope)
> ```

You may have to add extra logic into vcdcluster_conversion.go, please check that file to understand how other properties added in v1beta1 were dealt.

Also, please use debug setup with webhooks for your PR once your core proxy work is stabilized. Always test with webhooks enabled whenever you make changes to capiyaml.
___
*[`api/v1beta1/vcdcluster_types.go` line 86 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-N1opmKvVDIlUrqnSfZI-r1-86:-N1uG9lI3nl9ijatC09E:b-wb53tw) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/6da227c8a28bd4efed245e17d13cc7d4477a1269/api/v1beta1/vcdcluster_types.go#L86)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

there is a work from Yan I think which tries to keep status in sync with spec. So we need similar fields in status as well. @sahithi is that correct?
</blockquote></details>

Yes, Please populate these values in the Status section as well.
___
*[`controllers/cluster_scripts/cloud_init_control_plane.yaml` line 83 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-N1uGIb907t0-Wqd3siq:-N1uGIb907t0-Wqd3sir:b-wey5u7) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/6da227c8a28bd4efed245e17d13cc7d4477a1269/controllers/cluster_scripts/cloud_init_control_plane.yaml#L83)):*
> ```YAML
>       sed -i 's/::1/127.0.0.1/g' /etc/hosts || true
>     vmtoolsd --cmd "info-set guestinfo.postcustomization.networkconfiguration.status successful"
>     
> ```

As discussed Sakthi, I was trying to understand if this applies only to contained or if we want to support OS level proxy.
I think you already planning to make it OS-level proxy.

**Review by sakthisunda — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143)* status: 0 of 7 files reviewed, 5 unresolved discussions (waiting on @arunmk, @sahithi, and @sakthisunda)

___
*[`api/v1alpha4/zz_generated.conversion.go` line 111 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-N1uGaeHESjbR_c8nLoq:-N2N-TqIAV_RvWSFiVcb:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/6da227c8a28bd4efed245e17d13cc7d4477a1269/api/v1alpha4/zz_generated.conversion.go#L111)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

You may have to add extra logic into vcdcluster_conversion.go, please check that file to understand how other properties added in v1beta1 were dealt.

Also, please use debug setup with webhooks for your PR once your core proxy work is stabilized. Always test with webhooks enabled whenever you make changes to capiyaml.
</blockquote></details>

Done.
___
*[`api/v1beta1/vcdcluster_types.go` line 86 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-N1opmKvVDIlUrqnSfZI-r1-86:-N2N-UqPBstfIsxpBAxm:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/6da227c8a28bd4efed245e17d13cc7d4477a1269/api/v1beta1/vcdcluster_types.go#L86)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Yes, Please populate these values in the Status section as well.
</blockquote></details>

Done.
___
*[`controllers/cluster_scripts/cloud_init_control_plane.yaml` line 62 at r1](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-N1opmKvVDIlUrqnSfZK-r1-62:-N2N-XrQ2DkeG4VzxGTA:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/6da227c8a28bd4efed245e17d13cc7d4477a1269/controllers/cluster_scripts/cloud_init_control_plane.yaml#L62)):*
<details><summary><i>Previously, arunmk (Arun M. Krishnakumar) wrote…</i></summary><blockquote>

no need of this as mentioned above
</blockquote></details>

Done.

**Review by sahithi — APPROVED** [review]
> Sakthi, I don't yet see proxy config populated under status section.
> You need to do this in vcdcluster_controller.
:lgtm: after addressing the above comment

> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-:-N2RgZzj-NUJx6YzOOWC:bym920x)* status: 0 of 7 files reviewed, 6 unresolved discussions (waiting on @arunmk, @sahithi, and @sakthisunda)

___
*[`api/v1beta1/vcdcluster_types.go` line 113 at r4](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-N2RgNyrDGlmmA_kBZHY:-N2RgNyrDGlmmA_kBZHZ:b-gx1vs) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/79bdfe6fa21089089990b47c2ce456307b6b507c/api/v1beta1/vcdcluster_types.go#L113)):*
> ```Go
> 	UseAsManagementCluster bool `json:"useAsManagementCluster,omitempty"`
> 	// +optional
> 	ProxyConfig ProxyConfig `json:"proxyConfig,omitempty"`
> ```

Sakthi, I don't yet see proxy config populated under status section.
You need to do this in vcdcluster_controller.

**Review by sakthisunda — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143)* status: 0 of 8 files reviewed, 6 unresolved discussions (waiting on @arunmk, @sahithi, and @sakthisunda)

___
*[`api/v1beta1/vcdcluster_types.go` line 113 at r4](https://reviewable.io/reviews/vmware/cluster-api-provider-cloud-director/143#-N2RgNyrDGlmmA_kBZHY:-N2UqJLu4Bfpx74bt5KY:b-896fix) ([raw file](https://github.com/vmware/cluster-api-provider-cloud-director/blob/79bdfe6fa21089089990b47c2ce456307b6b507c/api/v1beta1/vcdcluster_types.go#L113)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Sakthi, I don't yet see proxy config populated under status section.
You need to do this in vcdcluster_controller.
</blockquote></details>

Done.

**Pull request merged**